File: mifosng-android/src/main/java/com/mifos/mifosxdroid/online/surveylist/SurveyListPresenter.java
package com.mifos.mifosxdroid.online.surveylist;

import com.mifos.api.datamanager.DataManagerSurveys;
import com.mifos.mifosxdroid.R;
import com.mifos.mifosxdroid.base.BasePresenter;
import com.mifos.objects.survey.QuestionDatas;
import com.mifos.objects.survey.ResponseDatas;
import com.mifos.objects.survey.Survey;
import com.mifos.utils.PrefManager;

import java.util.List;

import javax.inject.Inject;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.observers.DisposableObserver;
import io.reactivex.schedulers.Schedulers;


/**
 * Created by Rajan Maurya on 08/06/16.
 */
public class SurveyListPresenter extends BasePresenter<SurveyListMvpView> {

    private final DataManagerSurveys mDataManagerSurveys;
    private CompositeDisposable compositeDisposable;
    private List<Survey> mDbSurveyList;
    private List<Survey> mSyncSurveyList;

    @Inject
    public SurveyListPresenter(DataManagerSurveys dataManagerSurveys) {
        mDataManagerSurveys = dataManagerSurveys;
        compositeDisposable = new CompositeDisposable();
    }

    @Override
    public void attachView(SurveyListMvpView mvpView) {
        super.attachView(mvpView);
    }

    @Override
    public void detachView() {
        super.detachView();
        if (compositeDisposable != null) compositeDisposable.clear();
    }

    public void loadSurveyList() {
        checkViewAttached();
        getMvpView().showProgressbar(true);
        compositeDisposable.add(mDataManagerSurveys.getAllSurvey()
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeOn(Schedulers.io())
                .subscribeWith(new DisposableObserver<List<Survey>>() {
                    @Override
                    public void onComplete() {
                    }

                    @Override
                    public void onError(Throwable e) {
                        getMvpView().showProgressbar(false);
                        getMvpView().showFetchingError(R.string.failed_to_fetch_surveys_list);
                    }

                    @Override
                    public void onNext(List<Survey> surveys) {
                        mSyncSurveyList = surveys;
                        getMvpView().showProgressbar(false);
                        loadDatabaseSurveys();
                    }
                })
        );
    }

    public void loadDatabaseSurveys() {
        checkViewAttached();
        compositeDisposable.add(mDataManagerSurveys.getDatabaseSurveys()
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeOn(Schedulers.io())
                .subscribeWith(new DisposableObserver<List<Survey>>() {
                    @Override
                    public void onComplete() {
                        setAlreadySurveySyncStatus(mSyncSurveyList);
                        getMvpView().showAllSurvey(mSyncSurveyList);
                    }

                    @Override
                    public void onError(Throwable e) {
                        getMvpView().showFetchingError(R.string.failed_to_load_db_surveys);
                    }

                    @Override
                    public void onNext(List<Survey> surveyList) {
                        mDbSurveyList = surveyList;
                        if (PrefManager.getUserStatus() == 1) {
                            for (Survey survey : mSyncSurveyList) {
                                loadDatabaseQuestionDatas(survey.getId(), survey);
                            }
                        }
                    }
                })
        );
    }

    public void loadDatabaseQuestionDatas(int surveyId, final Survey survey) {
        checkViewAttached();
        compositeDisposable.add(mDataManagerSurveys.getDatabaseQuestionDatas(surveyId)
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeOn(Schedulers.io())
                .subscribeWith(new DisposableObserver<List<QuestionDatas>>() {
                    @Override
                    public void onComplete() {
                    }

                    @Override
                    public void onError(Throwable e) {
                        getMvpView().showFetchingError(R.string.failed_to_load_db_question_datas);
                    }

                    @Override
                    public void onNext(List<QuestionDatas> questionDatasList) {
                        for (QuestionDatas questionDatas : questionDatasList) {
                            loadDatabaseResponseDatas(questionDatas.getId(), questionDatas);
                        }
                        survey.setQuestionDatas(questionDatasList);
                    }
                })
        );
    }

    public void loadDatabaseResponseDatas(int questionId, final QuestionDatas questionDatas) {
        checkViewAttached();
        compositeDisposable.add(mDataManagerSurveys.getDatabaseResponseDatas(questionId)
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeOn(Schedulers.io())
                .subscribeWith(new DisposableObserver<List<ResponseDatas>>() {
                    @Override
                    public void onComplete() {
                    }

                    @Override
                    public void onError(Throwable e) {
                        getMvpView().showFetchingError(R.string.failed_to_load_db_response_datas);
                    }

                    @Override
                    public void onNext(List<ResponseDatas> responseDatas) {
                        questionDatas.setResponseDatas(responseDatas);
                    }
                })
        );
    }

    public void setAlreadySurveySyncStatus(List<Survey> surveys) {
        checkSurveyAlreadySyncedOrNot(surveys);
    }

    public void checkSurveyAlreadySyncedOrNot(List<Survey> surveys) {
        if (mDbSurveyList.size() != 0) {
            for (Survey dbSurvey : mDbSurveyList) {
                for (Survey syncSurvey : surveys) {
                    if (dbSurvey.getId() == syncSurvey.getId()) {
                        syncSurvey.setSync(true);
                        break;
                    }
                }
            }
        }
    }
}

File: mifosng-android/src/main/java/com/mifos/mifosxdroid/online/surveylist/SurveyListPresenter.kt
package com.mifos.mifosxdroid.online.surveylist

import com.mifos.api.datamanager.DataManagerSurveys
import com.mifos.mifosxdroid.R
import com.mifos.mifosxdroid.base.BasePresenter
import com.mifos.objects.survey.QuestionDatas
import com.mifos.objects.survey.ResponseDatas
import com.mifos.objects.survey.Survey
import com.mifos.utils.PrefManager
import rx.Subscriber
import rx.android.schedulers.AndroidSchedulers
import rx.schedulers.Schedulers
import rx.subscriptions.CompositeSubscription
import javax.inject.Inject

/**
 * Created by Rajan Maurya on 08/06/16.
 */
class SurveyListPresenter @Inject constructor(private val mDataManagerSurveys: DataManagerSurveys) :
    BasePresenter<SurveyListMvpView?>() {
    private val mSubscriptions: CompositeSubscription?
    private var mDbSurveyList: List<Survey>? = null
    private var mSyncSurveyList: List<Survey> ?= null

    override fun detachView() {
        super.detachView()
        mSubscriptions?.clear()
    }

    fun loadSurveyList() {
        checkViewAttached()
        mvpView!!.showProgressbar(true)
        mSubscriptions!!.add(
            mDataManagerSurveys.allSurvey
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeOn(Schedulers.io())
                .subscribe(object : Subscriber<List<Survey>>() {
                    override fun onCompleted() {}
                    override fun onError(e: Throwable) {
                        mvpView!!.showProgressbar(false)
                        mvpView!!.showFetchingError(R.string.failed_to_fetch_surveys_list)
                    }

                    override fun onNext(surveys: List<Survey>) {
                        mSyncSurveyList = surveys
                        mvpView!!.showProgressbar(false)
                        loadDatabaseSurveys()
                    }
                })
        )
    }

    fun loadDatabaseSurveys() {
        checkViewAttached()
        mSubscriptions!!.add(
            mDataManagerSurveys.databaseSurveys
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeOn(Schedulers.io())
                .subscribe(object : Subscriber<List<Survey>>() {
                    override fun onCompleted() {
                        mSyncSurveyList?.let { setAlreadySurveySyncStatus(it) }
                        mvpView!!.showAllSurvey(mSyncSurveyList)
                    }

                    override fun onError(e: Throwable) {
                        mvpView!!.showFetchingError(R.string.failed_to_load_db_surveys)
                    }

                    override fun onNext(surveyList: List<Survey>) {
                        mDbSurveyList = surveyList
                        if (PrefManager.userStatus == 1) {
                            for (survey in mSyncSurveyList!!) {
                                loadDatabaseQuestionData(survey!!.id, survey)
                            }
                        }
                    }
                })
        )
    }

    fun loadDatabaseQuestionData(surveyId: Int, survey: Survey?) {
        checkViewAttached()
        mSubscriptions!!.add(
            mDataManagerSurveys.getDatabaseQuestionData(surveyId)
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeOn(Schedulers.io())
                .subscribe(object : Subscriber<List<QuestionDatas>>() {
                    override fun onCompleted() {}
                    override fun onError(e: Throwable) {
                        mvpView!!.showFetchingError(R.string.failed_to_load_db_question_datas)
                    }

                    override fun onNext(questionDatasList: List<QuestionDatas>) {
                        for (questionDatas in questionDatasList) {
                            questionDatas.id?.let { loadDatabaseResponseDatas(it, questionDatas) }
                        }
                        survey!!.questionDatas = questionDatasList
                    }
                })
        )
    }

    fun loadDatabaseResponseDatas(questionId: Int, questionDatas: QuestionDatas) {
        checkViewAttached()
        mSubscriptions!!.add(
            mDataManagerSurveys.getDatabaseResponseDatas(questionId)
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeOn(Schedulers.io())
                .subscribe(object : Subscriber<List<ResponseDatas>>() {
                    override fun onCompleted() {}
                    override fun onError(e: Throwable) {
                        mvpView!!.showFetchingError(R.string.failed_to_load_db_response_datas)
                    }

                    override fun onNext(responseDatas: List<ResponseDatas>) {
                        questionDatas.responseDatas = responseDatas
                    }
                })
        )
    }

    fun setAlreadySurveySyncStatus(surveys: List<Survey>) {
        checkSurveyAlreadySyncedOrNot(surveys)
    }

    private fun checkSurveyAlreadySyncedOrNot(surveys: List<Survey>) {
        if (mDbSurveyList!!.isNotEmpty()) {
            for (dbSurvey in mDbSurveyList!!) {
                for (syncSurvey in surveys) {
                    if (dbSurvey.id == syncSurvey.id) {
                        syncSurvey.isSync = true
                        break
                    }
                }
            }
        }
    }

    init {
        mSubscriptions = CompositeSubscription()
    }
}